lua/hop/hint.lua

```diff
@@ -38,19 +42,248 @@ end
+          s = s:sub(0, cursor_pos[2] - start)
+        end
+
+        _,i = vim.regex('^\\s*'):match_str(s)
+        _,j = vim.regex('^\\s*\\(\\(\\w\\+\\)\\|\\([^[:blank:][:alnum:]_]\\+\\)\\)'):match_str(s)
+        j = j or s:len()
+
+        if i == j then
+          return 0, 0
+        else
+          if line_nr + 1 == cursor_pos[1] then
+            start = start + j
+          end
+          return i, j
+        end
+      else
+        return 0, 0
+      end
+    end
+  }
+end
+
+function M.by_find(c, only_to)
+  first_time = true
+  return {
+    oneshot = false,
+    match = function(s, line_nr, cursor_pos)
+      if line_nr + 1 == cursor_pos[1] then
+        if first_time then
+          if only_to then
+            line_start = cursor_pos[2] + 1
+          else
+            line_start = cursor_pos[2]
+          end
+          s = s:sub(line_start+2)
+        end
+        if c:match('%W') then
+          i = s:find('%' .. c)
+        else
+          i = s:find(c)
+        end
+        if i == nil then
+          return nil, nil
+        else
+          if only_to then
+            i = i - 1
+          end
+          if first_time then
+            first_time = false
+            return i + line_start, i + 2 + line_start
+          else
+            return i-1, i+1
+          end
+        end
+      end
+    end
+  }
+end
+
+function M.by_find_before(c, only_to)
+  start = 0
+  return {
+    oneshot = false,
+    match = function(s, line_nr, cursor_pos)
+      if line_nr + 1 == cursor_pos[1] then
+        if only_to then
+          line_start = cursor_pos[2] - 1
+        else
+          line_start = cursor_pos[2]
+        end
+        s = s:sub(0, line_start - start)
+        if c:match('%W') then
+          i = s:find('%' .. c)
+        else
+          i = s:find(c)
+        end
+
+        if i == nil then
+          return nil, nil
+        else
+          if only_to then
+            i = i + 1
+          end
+          start = start + i
+          return i-1, i
+        end
+      end
+    end
+  }
+end
 
 -- Line hint mode.
 --
 -- Used to tag the beginning of each lines with ihnts.
 M.by_line_start = {
   oneshot = true,
-  match = function(_)
+  match = function(_, _, _)
     return 0, 1, false
   end
 }
 
+M.by_line_down = {
+  oneshot = true,
+  match = function(s, line_nr, cursor_pos)
+    if line_nr >= cursor_pos[1] then
+      if s:len() == 0 then
+        line_max = s:len()
+      else
+        line_max = s:len() - 1
+      end
+      start = math.min(cursor_pos[2], line_max)
+      return start, start+1
+    else
+      return 0, 0
+    end
+  end
+}
+
+M.by_line_up = {
+  oneshot = true,
+  match = function(s, line_nr, cursor_pos)
+    if line_nr + 1 < cursor_pos[1] then
+      if s:len() == 0 then
+        line_max = s:len()
+      else
+        line_max = s:len() - 1
+      end
+      start = math.min(cursor_pos[2], line_max)
+      return start, start+1
+    else
+      return 0, 0
+    end
+  end
+}
+
```

Comment:
I do think you duplicated a lot of the logic of the already existing code instead of changing the inputs. For instance, `by_line_down` could be implemented without adding any new function and just adapting the input we are using — i.e. instead of using `top_line` and `bottom_line`, we could replace `top_line` by the cursor line.

This is true for, I guess, a lot of other commands you added. Some of them require fixing the match function, such as starting in the middle of a line (like what you did with all the variations of `f`, `F`, `t`, `t`, `w`, etc. However, same remark as above: if we use the exact same algorithms already implemented by providing an array of a single line, a lot of code you wrote there is non-needed.